pkg/typesystem/resolver.go

```diff
@@ -0,0 +1,79 @@
+package typesystem
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/karlseguin/ccache/v3"
+	"github.com/oklog/ulid/v2"
+	"github.com/openfga/openfga/pkg/storage"
+)
+
+const (
+	typesystemCacheTTL = 168 * time.Hour // 7 days
+)
+
+// TypesystemResolverFunc is a function that implementations can implement to provide lookup and
+// resolution of a Typesystem.
+type TypesystemResolverFunc func(ctx context.Context, storeID, modelID string) (*TypeSystem, error)
+
+// MemoizedTypesystemResolverFunc returns a TypesystemResolverFunc that either fetches the provided authorization
+// model (if provided) or looks up the latest authorizaiton model, and then it constructs a TypeSystem from
+// the resolved model. The type-system resolution is memoized so if another lookup of the same model occurs,
+// then the earlier TypeSystem that was constructed will be used.
+//
+// The memoized resolver function is safe for concurrent use.
+func MemoizedTypesystemResolverFunc(datastore storage.AuthorizationModelReadBackend) TypesystemResolverFunc {
+
+	cache := ccache.New(ccache.Configure[*TypeSystem]())
+
+	return func(ctx context.Context, storeID, modelID string) (*TypeSystem, error) {
+		ctx, span := tracer.Start(ctx, "MemoizedTypesystemResolverFunc")
+		defer span.End()
+
+		var err error
+
+		if modelID != "" {
+			if _, err := ulid.Parse(modelID); err != nil {
+				return nil, ErrModelNotFound
+			}
+		}
+
+		if modelID == "" {
+			if modelID, err = datastore.FindLatestAuthorizationModelID(ctx, storeID); err != nil {
+				if errors.Is(err, storage.ErrNotFound) {
+					return nil, ErrModelNotFound
+				}
+
+				return nil, fmt.Errorf("failed to FindLatestAuthorizationModelID: %w", err)
+			}
+		}
+
+		key := fmt.Sprintf("%s/%s", storeID, modelID)
+
+		item := cache.Get(key)
+		if item != nil {
+			return item.Value(), nil
+		}
+
+		model, err := datastore.ReadAuthorizationModel(ctx, storeID, modelID)
+		if err != nil {
+			if errors.Is(err, storage.ErrNotFound) {
+				return nil, ErrModelNotFound
+			}
+
+			return nil, fmt.Errorf("failed to ReadAuthorizationModel: %w", err)
+		}
+
+		typesys, err := NewAndValidate(ctx, model)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %v", ErrInvalidModel, err)
+		}
```

Comment:
If someone continuously passes an invalid auth model, we are not storing it anywhere in the cache and we'll try to fetch it from the DB every time 🤔

Reply:
Yeah, I purposefully chose not to cache invalid models because of the footprint, and if the model is invalid then the request can't be served successfully anyways, and so the SLO latency expectations around those cases aren't as tight tolerance as the alternative.

If you are referencing an invalid auth model and you experience a little more response latency, but you reduce memory footprint because you don't cache the invalid models, then that is a fine trade-off IMO.